pytorch_lightning/utilities/cli.py

```diff
@@ -334,16 +341,53 @@ def add_arguments_to_parser(self, parser: LightningArgumentParser) -> None:
             parser: The parser object to which arguments can be added
         """
 
-    def link_optimizers_and_lr_schedulers(self) -> None:
+    @staticmethod
+    def subcommands() -> Dict[str, Set[str]]:
+        """Defines the list of available subcommands and the arguments to skip."""
+        return {
+            "fit": {"model", "train_dataloaders", "train_dataloader", "val_dataloaders", "datamodule"},
+            "validate": {"model", "dataloaders", "val_dataloaders", "datamodule"},
+            "test": {"model", "dataloaders", "test_dataloaders", "datamodule"},
+            "predict": {"model", "dataloaders", "datamodule"},
+            "tune": {"model", "train_dataloaders", "train_dataloader", "val_dataloaders", "datamodule"},
+        }
+
+    def _add_subcommands(self, parser: LightningArgumentParser, **kwargs: Any) -> None:
+        """Adds subcommands to the input parser."""
+        parser_subcommands = parser.add_subcommands()
+        # the user might have passed a builder function
+        trainer_class = (
+            self.trainer_class if isinstance(self.trainer_class, type) else class_from_function(self.trainer_class)
+        )
+
+        for subcommand in self.subcommands():
```

Comment:
Do we need to register all parser if only 1 subcommand is being run ?

Reply:
Yes, otherwise setting configs (for more than one subcommand) before the subcommand wouldn't work

Also we don't know the subcommand before we parse the args (unless we peek at argv) and we cannot parse the args if we dont' register all

And we want all info to appear in the `--help` message